[JSC] add support for `Map.prototype.getOrInsert` et al

dcrousso opened the pull request:
#### 2b8f7267635745ac735406185b3f36348c5e82fc
<pre>
[JSC] add support for `Map.prototype.getOrInsert` et al
<a href="https://bugs.webkit.org/show_bug.cgi?id=282014">https://bugs.webkit.org/show_bug.cgi?id=282014</a>

Reviewed by Yusuke Suzuki.

It&apos;s an extremely common pattern to check for the existence of a key in a `Map` (or `WeakMap`) before adding a new entry.

```js
if (map.has(key))
    return map.get(key);
map.set(key, value);
return value;
```

The primary benefit of of `getOrInsert` is that it avoids having to re-hash the key.

The additional benefit of `getOrInsertComputed` beyond the above is that it doesn&apos;t have to compute the value unless the key is not present.

For both `Map` and `WeakMap` it was necesasry to create additional helper methods that returned the hash and/or existing entry.

Spec: &lt;<a href="https://tc39.es/proposal-upsert/">https://tc39.es/proposal-upsert/</a>&gt;
Proposal: &lt;<a href="https://github.com/tc39/proposal-upsert">https://github.com/tc39/proposal-upsert</a>&gt;

* Source/JavaScriptCore/runtime/MapPrototype.cpp:
(JSC::MapPrototype::finishCreation):
(JSC::mapProtoFuncGetOrInsert): Added.
(JSC::mapProtoFuncGetOrInsertComputed): Added.
* Source/JavaScriptCore/runtime/OrderedHashTable.h:
(JSC::OrderedHashMap::getOrInsert): Added.
* Source/JavaScriptCore/runtime/OrderedHashTableHelper.h:
(JSC::OrderedHashTableHelper::addImpl):

* Source/JavaScriptCore/runtime/WeakMapPrototype.cpp:
(JSC::WeakMapPrototype::finishCreation):
(JSC::protoFuncWeakMapGetOrInsert): Added.
(JSC::protoFuncWeakMapGetOrInsertComputed): Added.
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
(JSC::WeakMapImpl::getBucket): Added.
(JSC::WeakMapImpl::findBucketIndex): Added.
(JSC::WeakMapImpl::findBucket):
(JSC::WeakMapImpl::findBucketIndexAlreadyHashed): Added.
(JSC::WeakMapImpl::findBucketAlreadyHashed): Deleted.
* Source/JavaScriptCore/runtime/WeakMapImplInlines.h:
(JSC::WeakMapImpl&lt;WeakMapBucket&gt;::addBucket): Added.

* Source/JavaScriptCore/runtime/OptionsList.h:

* JSTests/stress/map-getOrInsert.js: Added.
* JSTests/stress/map-getOrInsertComputed.js: Added.
* JSTests/stress/weakmap-getOrInsert.js: Added.
* JSTests/stress/weakmap-getOrInsertComputed.js: Added.

Canonical link: <a href="https://commits.webkit.org/286377@main">https://commits.webkit.org/286377@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/250aa61239c3fddc2e3fc73dbb1ed1d2c18dbf2a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75821 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/54850 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28718 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80313 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27087 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/63992 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3144 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59465 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17622 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78888 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49331 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65119 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39825 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46730 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22602 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25414 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/68998 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67847 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22940 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81782 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/75097 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3190 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2000 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67688 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3344 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65086 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66993 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16699 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10930 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9058 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/97351 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3140 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21284 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3161 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4099 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3168 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/MapPrototype.cpp:
```diff
@@ -176,6 +186,56 @@ JSC_DEFINE_HOST_FUNCTION(mapProtoFuncSet, (JSGlobalObject* globalObject, CallFra
+    if (!valueCallback.isCallable())
+        return throwVMTypeError(globalObject, scope, "Map.prototype.getOrInsertComputed requires the callback argument to be callable."_s);
+
+    return JSValue::encode(map->getOrInsert(globalObject, key, [&] {
```
Wrap it with `RELEASE_AND_RETURN(scope, ...)`

Constellation commented on Source/JavaScriptCore/runtime/MapPrototype.cpp:
```diff
@@ -176,6 +186,56 @@ JSC_DEFINE_HOST_FUNCTION(mapProtoFuncSet, (JSGlobalObject* globalObject, CallFra
+
+    return JSValue::encode(map->getOrInsert(globalObject, key, [&] {
+        return callFrame->argument(1);
+    }));
```
Ditto.

dcrousso committed:
[JSC] add support for `Map.prototype.getOrInsert` et al
https://bugs.webkit.org/show_bug.cgi?id=282014

Reviewed by Yusuke Suzuki.

It's an extremely common pattern to check for the existence of a key in a `Map` (or `WeakMap`) before adding a new entry.

```js
if (map.has(key))
    return map.get(key);
map.set(key, value);
return value;
```

The primary benefit of of `getOrInsert` is that it avoids having to re-hash the key.

The additional benefit of `getOrInsertComputed` beyond the above is that it doesn't have to compute the value unless the key is not present.

For both `Map` and `WeakMap` it was necesasry to create additional helper methods that returned the hash and/or existing entry.

Spec: <https://tc39.es/proposal-upsert/>
Proposal: <https://github.com/tc39/proposal-upsert>

* Source/JavaScriptCore/runtime/MapPrototype.cpp:
(JSC::MapPrototype::finishCreation):
(JSC::mapProtoFuncGetOrInsert): Added.
(JSC::mapProtoFuncGetOrInsertComputed): Added.
* Source/JavaScriptCore/runtime/OrderedHashTable.h:
(JSC::OrderedHashMap::getOrInsert): Added.
* Source/JavaScriptCore/runtime/OrderedHashTableHelper.h:
(JSC::OrderedHashTableHelper::addImpl):

* Source/JavaScriptCore/runtime/WeakMapPrototype.cpp:
(JSC::WeakMapPrototype::finishCreation):
(JSC::protoFuncWeakMapGetOrInsert): Added.
(JSC::protoFuncWeakMapGetOrInsertComputed): Added.
* Source/JavaScriptCore/runtime/WeakMapImpl.h:
(JSC::WeakMapImpl::getBucket): Added.
(JSC::WeakMapImpl::findBucketIndex): Added.
(JSC::WeakMapImpl::findBucket):
(JSC::WeakMapImpl::findBucketIndexAlreadyHashed): Added.
(JSC::WeakMapImpl::findBucketAlreadyHashed): Deleted.
* Source/JavaScriptCore/runtime/WeakMapImplInlines.h:
(JSC::WeakMapImpl<WeakMapBucket>::addBucket): Added.

* Source/JavaScriptCore/runtime/OptionsList.h:

* JSTests/stress/map-getOrInsert.js: Added.
* JSTests/stress/map-getOrInsertComputed.js: Added.
* JSTests/stress/weakmap-getOrInsert.js: Added.
* JSTests/stress/weakmap-getOrInsertComputed.js: Added.

Canonical link: https://commits.webkit.org/286377@main
